File: app/src/main/java/org/tvheadend/tvhclient/features/shared/tasks/HtmlFileLoaderTask.java
package org.tvheadend.tvhclient.features.shared.tasks;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.utils.MiscUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.lang.ref.WeakReference;
import java.util.Locale;

import androidx.annotation.NonNull;
import timber.log.Timber;

public class HtmlFileLoaderTask extends AsyncTask<Void, Void, String> {

    private final String file;
    private final String defaultLocale;
    private final WeakReference<Context> context;
    private final FileLoaderCallback callback;

    public HtmlFileLoaderTask(Context context, String file, String defaultLocale, @NonNull FileLoaderCallback callback) {
        this.context = new WeakReference<>(context);
        this.file = file;
        this.defaultLocale = defaultLocale;
        this.callback = callback;
    }

    @Override
    protected String doInBackground(Void... voids) {
        Context ctx = context.get();
        if (ctx != null) {
            return loadHtmlFromFile(ctx, file, defaultLocale);
        }
        return null;
    }

    @Override
    protected void onPostExecute(String content) {
        callback.notify(content);
    }

    private String loadHtmlFromFile(Context context, String filename, String defaultLocale) {
        // Create the string that is later used to display the HTML page.
        // The string contains all feature information and HTML tags.
        // Depending on the theme the correct style sheet will be loaded
        // from the asset folder.
        StringBuilder sb = new StringBuilder();
        sb.append("<html><head>");
        if (MiscUtils.getThemeId(context) == R.style.CustomTheme_Light) {
            sb.append("<link href=\"html/styles_light.css\" type=\"text/css\" rel=\"stylesheet\"/>");
        } else {
            sb.append("<link href=\"html/styles_dark.css\" type=\"text/css\" rel=\"stylesheet\"/>");
        }
        sb.append("</head><body>");

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        final Locale current = context.getResources().getConfiguration().locale;
        final String locale = prefs.getString("languagePref", current.getLanguage()).substring(0, 2);
        final String htmlFile = "html/" + filename + "_" + locale.substring(0, 2) + ".html";
        final String defaultHtmlFile = "html/" + filename + "_" + defaultLocale + ".html";

        // Open the HTML file of the defined language. This is determined by
        // the defaultLocale. If the file doesn't exist, open the default (English)
        InputStream is = null;
        try {
            is = context.getAssets().open(htmlFile);
        } catch (IOException ex1) {
            Timber.e("Could not open file " + htmlFile);
        }
        if (is == null) {
            try {
                is = context.getAssets().open(defaultHtmlFile);
            } catch (IOException ex2) {
                Timber.e("Could not open default file " + defaultHtmlFile);
            }
        }

        // Try to parse the HTML contents from the given asset file. It
        // contains the feature description with the required HTML tags.
        if (is != null) {
            try {
                String htmlData;
                BufferedReader in = new BufferedReader(new InputStreamReader(is, "utf-8"));
                while ((htmlData = in.readLine()) != null) {
                    sb.append(htmlData);
                }
                in.close();

            } catch (UnsupportedEncodingException uee) {
                Timber.e("Could not create buffered reader, unsupported encoding");
                sb.append("Error parsing feature list");
            } catch (IOException ioe) {
                Timber.e("Error while reading contents from input stream or closing it");
                sb.append("Error parsing feature list");
            }
            try {
                is.close();
            } catch (IOException e) {
                Timber.e("Error closing input stream");
            }
        }

        // Add the closing HTML tags and load show the page
        sb.append("</body></html>");
        Timber.d("Done loading file");
        return sb.toString();
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/information/HtmlFileLoaderTask.kt
package org.tvheadend.tvhclient.ui.features.information

import android.content.Context
import android.os.AsyncTask
import android.os.Build
import androidx.preference.PreferenceManager
import org.tvheadend.tvhclient.ui.common.getLocale
import timber.log.Timber
import java.io.*
import java.lang.ref.WeakReference
import java.nio.charset.StandardCharsets

class HtmlFileLoaderTask(context: Context, private val file: String, private val defaultLocale: String, private val callback: Listener) : AsyncTask<Void, Void, String>() {
    private val context: WeakReference<Context> = WeakReference(context)

    override fun doInBackground(vararg voids: Void): String? {
        val ctx = context.get()
        return if (ctx != null) {
            loadHtmlFromFile(ctx, file)
        } else null
    }

    override fun onPostExecute(content: String) {
        callback.onFileContentsLoaded(content)
    }

    private fun loadHtmlFromFile(context: Context, filename: String): String {

        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val languageCode = prefs.getString("language", getLocale(context).language)!!.substring(0, 2)
        val htmlFile = "html/" + filename + "_" + languageCode.substring(0, 2) + ".html"
        val defaultHtmlFile = "html/" + filename + "_" + defaultLocale + ".html"

        // Open the HTML file of the defined language. This is determined by
        // the defaultLocale. If the file doesn't exist, open the default (English)
        var inputStream: InputStream? = null
        try {
            inputStream = context.assets.open(htmlFile)
        } catch (e: IOException) {
            Timber.e(e, "Could not open file $htmlFile")
        }

        if (inputStream == null) {
            try {
                inputStream = context.assets.open(defaultHtmlFile)
            } catch (e: IOException) {
                Timber.e(e, "Could not open default file $defaultHtmlFile")
            }
        }

        val sb = StringBuilder()
        // Try to parse the HTML contents from the given asset file. It
        // contains the feature description with the required HTML tags.
        if (inputStream != null) {
            try {
                val reader: BufferedReader = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8))
                } else {
                    BufferedReader(InputStreamReader(inputStream))
                }
                var htmlData: String
                while (true) {
                    htmlData = reader.readLine() ?: break
                    sb.append(htmlData)
                }
                reader.close()

            } catch (e: UnsupportedEncodingException) {
                Timber.e(e, "Could not create buffered reader, unsupported encoding")
            } catch (e: IOException) {
                Timber.e(e, "Error while reading contents from input stream or closing it")
            }

            try {
                inputStream.close()
            } catch (e: IOException) {
                Timber.e(e, "Error closing input stream")
            }

        }

        // Add the closing HTML tags and load show the page
        Timber.d("Done loading file")
        return sb.toString()
    }

    interface Listener {
        fun onFileContentsLoaded(fileContent: String)
    }
}
